### ISSUE TITLE
How to pass in Retry Option Settings when not using Dependency Injection (DI)?

### ISSUE THREAD

**Issue by kbdavis07** [ChangemakerStudios/GotenbergSharpApiClient#24]
I have the following Set-up

```
AppSettings

  "GotenbergSharpClient": {
    "ServiceUrl": "http://localhost:3000",
    "HealthCheckUrl": "http://localhost:3000/ping",
    "RetryPolicy": {
      "Enabled": true,
      "RetryCount": 4,
      "BackoffPower": 1.5,
      "LoggingEnabled": true
    }
  }
```

```
public void ConfigureServices(IServiceCollection services)
{
	.....
    services.AddOptions<GotenbergSharpClientOptions>()
	        .Bind(Configuration.GetSection(nameof(GotenbergSharpClient)));
    services.AddGotenbergSharpClient();
	.....    
}
```

**Question 1:**

I notice when I have this

`var test = Configuration.GetSection(nameof(GotenbergSharpClient));`

It is always null? 

So not sure if the App settings is getting passed on to the GotenbergSharpClient or not?

**Question 2:** 

When newing up the GotenbergSharpClient

` _Instance = new GotenbergSharpClient(Settings.Current.PDFServer);`

How does the Retry options get passed to it?

```
"RetryPolicy": {
      "Enabled": true,
      "RetryCount": 4,
      "BackoffPower": 1.5,
      "LoggingEnabled": true
    }
```

Not 100% sure when newing it up if the Retry Policies gets passed in to it?

I am thinking and hoping by looking at this

```
public static IAsyncPolicy<HttpResponseMessage> CreatePolicyFromSettings(IServiceProvider sp,
            HttpRequestMessage request)
```

it seems like the policy is created on HttpResponseMessage and it gets the settings from the service provider and since that
is provided in Startup.cs that the retry settings should be passed in regardless of how the Client is used?

**Question 3:**

How do I check to make sure that the retry settings are being passed into the Gotenberg Service correctly?

**Question 4:**

What is the best way to Unit test Gotenberg Server Errors which results in HttpResonse errors, to trigger the retry policy to kick in?

Thanks 👍

**Comment by chrismo111** [comment]
### Question 2:
GotenbergSharpClient was built as a “[typed client](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#how-to-use-typed-clients-with-ihttpclientfactory)” for use with dependency injection.  I'll make that more clear in the ReadMe.

The constructor overloads for [string and Uri](https://github.com/ChangemakerStudios/GotenbergSharpApiClient/blob/develop/lib/GotenbergSharpClient.cs#L37)  were added to simplify the [LinqPad demo code samples](https://github.com/ChangemakerStudios/GotenbergSharpApiClient/tree/develop/linqpad).   I should've decorated them with the [Obsolete attribute ](https://docs.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=netcore-3.1)or [EditorBrowsable](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.editorbrowsableattribute?view=netcore-3.1)(EditorBrowsableState.Never).   The overload that accepts an instance of HttpClient shouldn't be used either (I need to remove the PublicApi attribute from that).  

[GotenbergSharpClient](https://github.com/ChangemakerStudios/GotenbergSharpApiClient/blob/develop/lib/GotenbergSharpClient.cs) and its inner HttpClient are [created and managed by HttpClientFactory behind the scenes](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#how-to-use-typed-clients-with-ihttpclientfactory) when you call [my  AddGotenbergSharpClient service collection extension](https://github.com/ChangemakerStudios/GotenbergSharpApiClient/blob/develop/lib/Extensions/TypedClientServiceCollectionExtensions.cs#L29) which in turn calls [AddHttpClient](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientfactoryservicecollectionextensions.addhttpclient?view=dotnet-plat-ext-3.1) and [AddTypedClient](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientbuilderextensions.addtypedclient?view=dotnet-plat-ext-3.1). 

Take a look at these resources that outline all the benefits of DI/HttpClientFactory with Typed Clients:

- [3 ways to use HTTPClientFactory in ASP.NET Core 2.1](https://www.talkingdotnet.com/3-ways-to-use-httpclientfactory-in-asp-net-core-2-1/)
- [HttpClientFactory in ASP.NET CORE 2.1, Part 1 An Intro](https://www.stevejgordon.co.uk/introduction-to-httpclientfactory-aspnetcore)
- [Part 2, Defining Named & Typed Clients](https://www.stevejgordon.co.uk/httpclientfactory-named-typed-clients-aspnetcore)

### Question 1:
To see the options set after the container is built just resolve them as IOptions through a service provider. E.G. `sp.GetRequiredService<IOptions<GotenbergSharpClientOptions>>().Value`

You could also get them in Startup like this:
```
var test = Configuration.GetSection(nameof(GotenbergSharpClient)).Get<GotenbergSharpClientOptions>();
```
When calling Configuration.GetSection alone the test.Value property is supposed to be null by design. From the [MS documentation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1): 

> "When GetSection returns a matching section, Value isn't populated. A Key and Path are returned when the section exists."

  See a discussion here: [https://stackoverflow.com/q/46017593](https://stackoverflow.com/q/46017593)

**Comment by kbdavis07** [comment]
@chrismo111 Thanks for your reply 👍 

Current situation I am in is we are using heavily static classes, project was using WebForms/WCF and ported to .Net Core, but slowly converting those to services that uses DI, but for PDF need to use it now in our static pdf helper for the time being.

So when used in a static class the Retry policy does not kick in then?

Is there a way to use a service locator in static class to get it to work?

Also still have

Question 4:

What is the best way to Unit test Gotenberg Server Errors which results in HttpResonse errors, to trigger the retry policy to kick in?

I also need to create Unit test that would cause the Retry to trigger.


Thanks again 👍

**Comment by chrismo111** [comment]
If you're willing to use a service locator that'll work. 

You'll still need to call services.AddGotenbergSharpClient to have your settings applied. When the service locator resolves an instance it'll be configured with those settings.  

YMMV but I whipped together a little proof of concept where everything worked, including retries, in both netcore app2 and 3x.   

Within a static class I retrieved the client like this:
 `var sharpClient = ServiceLocator.ServiceProvider.GetService<GotenbergSharpClient>(); `

Testing the retry policy? 
I'd start by looking at how Polly Tests them & their documentation:

-  [AsyncDemo02_WaitAndRetryNTimes.cs](https://github.com/App-vNext/Polly-Samples/blob/master/PollyTestClient/Samples/AsyncDemo02_WaitAndRetryNTimes.cs)
-  [Unit testing with Polly with examples](https://github.com/App-vNext/Polly/wiki/Unit-testing-with-Polly---with-examples)
